015-kafka-topic-encryption.md

```diff
@@ -61,17 +60,56 @@ Encrypter-decrypters require a means to lookup encryption requirements and param
+
+To illuminate the essential details of message encryption, an example is stepped through.
+The encryption module is embedded in a proxy and the proxy, as an intermediary, terminates all connections to and from the broker, placing it in the position of intercepting all Kafka traffic. The message handling component examines all messages to identify the apikey (i.e., the Kafka message type). All message types besides Produce and Fetch are passed immediately onto the broker without any processing. Produce requests are deserialized and, using the message handling component, further inspected to identify the topic for which records are destined.
+If the topic matches a topic in the encryption policy, the contents of the _value_ field of each record is encrypted and replaced with the corresponding ciphertext. Metadata about the encryption (e.g., key, nonce) is added to the record's headers in order that information necessary for later decrypting the message (also referred to as encryption metadata) is available.
+
+As mentioned, encryption metadata is currently stored in record headers. Kafka record headers are optional name value/pairs stored in an array associated with the record. Record header names are not required to be unique, thus name "collisions" are possible. Ordering and versioning are used to avoid problems arising from using header names which coincidentally equal names chosen by a Kafka client. Ordering means that the record header is rewritten such that the encryption-related headers
+occur first in the array. The first header always contains a version from which the exact
+metadata field names can be derived. The decryption algorithm therefore can discern between
+those headers of the encryption module and those of the client. Further, the version field
+is designed to handle future migration and backwards compatibility should changes, such as format or encryption options, in the semantics occur.
+
+Although not part of our reference implementation, the encryption of header values is a conceivable extension. In this case, policy would be extended to indicate that specific or all client header
+values be encrypted. Since record headers are already processed during encryption, and the algorithm can discriminate between encryption and client headers, it should be just an incremental change to logic to support encryption of header data as well. The same applies to the encryption of record header names although we recommend further investigating the implications of encrypting header names.
+
```

Comment:
Should we add a sentence saying whether we want this capability to be part of the proposal or not? Like this, it's not clear if we want to implement or not.

Reply:
Good point. I am not sure if this is a hard requirment, so did not want to volunteer to do this. I mainly wanted to  address the topic as a result of the strimzi PR discussion. The way it is worded indeed comes off a bit wishy-washy.

What's our consensus?

Reply:
I think we should state it will NOT be part of the initial work and if someone feels strongly about it they will say.